packages/client/src/DataCommonsClient.ts

```diff
@@ -0,0 +1,480 @@
+   * @param targetDate date string
+   */
+  private getClosestObservationToDate(
+    observations: Observation[],
+    targetDate: string
+  ): Observation | undefined {
+    const index = _.sortedIndexBy(
+      observations,
+      { value: 0, date: targetDate },
+      (o) => o.date
+    );
+    return observations[index];
+  }
+
+  /**
+   * Returns all entity DCIDs found in the given PointApiResponse
+   * @param pointApiResponse
+   * @returns entity DCIDs
+   */
+  private getEntityDcidsFromPointApiResponse(
+    pointApiResponse: PointApiResponse
+  ): string[] {
+    const allEntityDcids = new Set<string>();
+    Object.keys(pointApiResponse.data).forEach((variableDcid) => {
+      Object.keys(pointApiResponse.data[variableDcid]).forEach((entityDcid) => {
+        allEntityDcids.add(entityDcid);
+      });
+    });
+    return Array.from(allEntityDcids);
+  }
+
+  /**
+   * Enriches PointApiResponse and converts response into a list of `DataRow`s
+   * @param entityDcids Entity DCIDs
+   * @param variableDcids Variable DCIDs
+   * @param pointApiResponse Entity/variable observations
+   * @param entityPropValues Additional entity properties to fetch
+   * @param variablePropValues Additional variable properties to fetch
+   * @param populationObservations Population observations for our list of entities for per-capita calculations
+   * @returns data rows
+   */
+  private getDataRowsFromObservations(
+    entityDcids: string[],
+    variableDcids: string[],
+    pointApiResponse: PointApiResponse,
+    entityPropValues: NodePropValues,
+    variablePropValues: NodePropValues,
+    populationObservations: SeriesApiResponse
+  ) {
+    const dataRows: DataRow[] = [];
+    entityDcids.forEach((entityDcid) => {
+      const row: DataRow = {
+        "entity.dcid": entityDcid,
+      };
+      Object.keys(entityPropValues).forEach((entityProp) => {
+        row[`entity.${entityProp}`] = entityPropValues[entityProp][entityDcid];
+      });
+      variableDcids.forEach((variableDcid) => {
+        const observation =
+          pointApiResponse.data[variableDcid][entityDcid] || {};
+        const facet = _.get(
+          pointApiResponse.facets,
+          observation.facet || "",
+          {}
+        );
+        row[`${variableDcid}.value`] = _.get(observation, "value", null);
+        row[`${variableDcid}.date`] = _.get(observation, "date", null);
+        row[`${variableDcid}.unit`] = _.get(facet, "unit", null);
+        row[`${variableDcid}.unitDisplayName`] = _.get(
+          observation,
+          "unitDisplayName",
+          _.get(facet, "unitDisplayName", null)
+        );
+        Object.keys(variablePropValues).forEach((variableProp) => {
+          row[`${variableDcid}.${variableProp}`] =
+            variablePropValues[variableProp][variableDcid];
+        });
+        // Set per-capita data
+        if (TOTAL_POPULATION_VARIABLE in populationObservations.data) {
+          const series =
```

Comment:
also need to check that entityDcid is in `populationObservations.data[TOTAL_POPULATION_VARIABLE]`

Reply:
Added!